### ISSUE TITLE
JUnitPlatformTestTree does not add dynamic containers to test plan

### ISSUE THREAD

**Issue by mfulton26** [junit-team/junit5#366]
I'm trying to register dynamic tests in a hierarchy but when I run from my IDE via `RunWith(JUnit5.class)` I get `org.junit.gen5.commons.util.PreconditionViolationException: No TestIdentifier with unique ID [$uniqueId] has been added to this TestPlan`.

I've put together a `DemoTestEngine` to reproduce the issue which you can simply run via your IDE like you'd run any JUnit 4 test after registering it as a test engine.

From a quick look at the code it appears that `JUnit5RunnerListener` does not add dynamic containers to its test plan so creating a hierarchy of dynamic tests causes the precondition to fail.

```
Internal Error occured.
org.junit.gen5.commons.util.PreconditionViolationException: No TestIdentifier with unique ID [[engine:demo]/[engine:demo]/[container:1]] has been added to this TestPlan.
	at org.junit.gen5.commons.util.Preconditions.condition(Preconditions.java:161)
	at org.junit.gen5.launcher.TestPlan.getTestIdentifier(TestPlan.java:162)
	at org.junit.gen5.junit4.runner.JUnit5TestTree.addDynamicDescription(JUnit5TestTree.java:61)
	at org.junit.gen5.junit4.runner.JUnit5RunnerListener.dynamicTestRegistered(JUnit5RunnerListener.java:41)
	at org.junit.gen5.launcher.main.TestExecutionListenerRegistry$CompositeTestExecutionListener.lambda$dynamicTestRegistered$22(TestExecutionListenerRegistry.java:48)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at org.junit.gen5.launcher.main.TestExecutionListenerRegistry.notifyTestExecutionListeners(TestExecutionListenerRegistry.java:37)
	at org.junit.gen5.launcher.main.TestExecutionListenerRegistry.access$100(TestExecutionListenerRegistry.java:26)
	at org.junit.gen5.launcher.main.TestExecutionListenerRegistry$CompositeTestExecutionListener.dynamicTestRegistered(TestExecutionListenerRegistry.java:48)
	at org.junit.gen5.launcher.main.ExecutionListenerAdapter.dynamicTestRegistered(ExecutionListenerAdapter.java:41)
	at DemoTestEngine.execute(DemoTestEngine.java:32)
	at org.junit.gen5.launcher.main.DefaultLauncher.execute(DefaultLauncher.java:124)
	at org.junit.gen5.launcher.main.DefaultLauncher.execute(DefaultLauncher.java:84)
	at org.junit.gen5.junit4.runner.JUnit5.run(JUnit5.java:107)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:119)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:42)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
```

```java
@RunWith(JUnit5.class)
public class DemoTestEngine implements TestEngine {
    @Override
    public String getId() {
        return "demo";
    }

    @Override
    public TestDescriptor discover(EngineDiscoveryRequest discoveryRequest, UniqueId uniqueId) {
        EngineDescriptor engineDescriptor = new EngineDescriptor(uniqueId.append("engine", "demo"), "demo");
        engineDescriptor.addChild(new RootTestDescriptor(engineDescriptor.getUniqueId().append("root", "0"), "0"));
        return engineDescriptor;
    }

    @Override
    public void execute(ExecutionRequest request) {
        EngineExecutionListener engineExecutionListener = request.getEngineExecutionListener();
        TestDescriptor root = request.getRootTestDescriptor();
        TestDescriptor child = new TestTestDescriptor(root.getUniqueId().append("container", "1"), "1");
        root.addChild(child);
        engineExecutionListener.dynamicTestRegistered(child);
        engineExecutionListener.executionStarted(child);
        engineExecutionListener.executionFinished(child, TestExecutionResult.successful());
        TestDescriptor grandchild = new TestTestDescriptor(child.getUniqueId().append("test", "2"), "2");
        child.addChild(grandchild);
        engineExecutionListener.dynamicTestRegistered(grandchild);
        engineExecutionListener.executionStarted(grandchild);
        engineExecutionListener.executionFinished(grandchild, TestExecutionResult.successful());
    }

    abstract static class AbstractDisplayNamedTestDescriptor extends AbstractTestDescriptor {
        private final String displayName;

        AbstractDisplayNamedTestDescriptor(UniqueId uniqueId, String displayName) {
            super(uniqueId);
            this.displayName = displayName;
        }

        @Override
        public String getDisplayName() {
            return displayName;
        }
    }

    static class RootTestDescriptor extends AbstractDisplayNamedTestDescriptor {
        RootTestDescriptor(UniqueId uniqueId, String displayName) {
            super(uniqueId,displayName);
        }

        @Override
        public boolean isContainer() {
            return true;
        }

        @Override
        public boolean isTest() {
            return false;
        }

        @Override
        public boolean hasTests() {
            return true;
        }
    }

    static class ContainerTestDescriptor extends AbstractDisplayNamedTestDescriptor {
        ContainerTestDescriptor(UniqueId uniqueId, String displayName) {
            super(uniqueId, displayName);
        }

        @Override
        public boolean isContainer() {
            return true;
        }

        @Override
        public boolean isTest() {
            return false;
        }
    }

    static class TestTestDescriptor extends AbstractDisplayNamedTestDescriptor {
        TestTestDescriptor(UniqueId uniqueId, String displayName) {
            super(uniqueId, displayName);
        }

        @Override
        public boolean isContainer() {
            return false;
        }

        @Override
        public boolean isTest() {
            return true;
        }
    }
}
```

**Comment by sbrannen** [comment]
Thanks for raising the issue.

Your analysis is correct: it is not currently possible to add a dynamic `TestDescriptor` for a _container_. 

Your example engine is based on an older version of the code base. `JUnit5RunnerListener` is now `JUnitPlatformRunnerListener`.

In any case, the actual problem is in `JUnitPlatformTestTree.addDynamicDescription()`, and I will soon be committing a fix for that.

**Comment by mfulton26** [comment]
Excellent! Thanks!

**Comment by sbrannen** [comment]
If you want to continue your work against the current state of `master`, the following updated version of your demo engine will work (once I submit the fix).

```java
import org.junit.platform.engine.EngineDiscoveryRequest;
import org.junit.platform.engine.EngineExecutionListener;
import org.junit.platform.engine.ExecutionRequest;
import org.junit.platform.engine.TestDescriptor;
import org.junit.platform.engine.TestEngine;
import org.junit.platform.engine.TestExecutionResult;
import org.junit.platform.engine.UniqueId;
import org.junit.platform.engine.support.descriptor.AbstractTestDescriptor;
import org.junit.platform.engine.support.descriptor.EngineDescriptor;
import org.junit.platform.runner.JUnitPlatform;
import org.junit.runner.RunWith;

@RunWith(JUnitPlatform.class)
public class DemoTestEngine implements TestEngine {

	private UniqueId rootUid;

	@Override
	public String getId() {
		return "demo";
	}

	@Override
	public TestDescriptor discover(EngineDiscoveryRequest discoveryRequest, UniqueId uniqueId) {
		EngineDescriptor engineDescriptor = new EngineDescriptor(uniqueId, "demo");
		RootTestDescriptor root = new RootTestDescriptor(uniqueId.append("root", "0"), "root #0");
		engineDescriptor.addChild(root);
		this.rootUid = root.getUniqueId();
		return engineDescriptor;
	}

	@Override
	public void execute(ExecutionRequest request) {
		EngineExecutionListener engineExecutionListener = request.getEngineExecutionListener();
		TestDescriptor engineDescriptor = request.getRootTestDescriptor();
		TestDescriptor root = engineDescriptor.findByUniqueId(this.rootUid).get();

		TestDescriptor container = new ContainerTestDescriptor(this.rootUid.append("container", "1"), "container #1");
		root.addChild(container);

		engineExecutionListener.dynamicTestRegistered(container);
		engineExecutionListener.executionStarted(container);
		engineExecutionListener.executionFinished(container, TestExecutionResult.successful());

		UniqueId containerUid = container.getUniqueId();

		TestDescriptor dynamicTest1 = new TestTestDescriptor(containerUid.append("test", "1"), "dynamic test #1");
		container.addChild(dynamicTest1);
		engineExecutionListener.dynamicTestRegistered(dynamicTest1);
		engineExecutionListener.executionStarted(dynamicTest1);
		engineExecutionListener.executionFinished(dynamicTest1, TestExecutionResult.successful());

		TestDescriptor dynamicTest2 = new TestTestDescriptor(containerUid.append("test", "2"), "dynamic test #2");
		container.addChild(dynamicTest2);
		engineExecutionListener.dynamicTestRegistered(dynamicTest2);
		engineExecutionListener.executionStarted(dynamicTest2);
		engineExecutionListener.executionFinished(dynamicTest2, TestExecutionResult.successful());
	}

	static class RootTestDescriptor extends AbstractTestDescriptor {

		RootTestDescriptor(UniqueId uniqueId, String displayName) {
			super(uniqueId, displayName);
		}

		@Override
		public boolean isContainer() {
			return true;
		}

		@Override
		public boolean isTest() {
			return false;
		}

		@Override
		public boolean hasTests() {
			return true;
		}
	}

	static class ContainerTestDescriptor extends AbstractTestDescriptor {

		ContainerTestDescriptor(UniqueId uniqueId, String displayName) {
			super(uniqueId, displayName);
		}

		@Override
		public boolean isContainer() {
			return true;
		}

		@Override
		public boolean isTest() {
			return false;
		}
	}

	static class TestTestDescriptor extends AbstractTestDescriptor {

		TestTestDescriptor(UniqueId uniqueId, String displayName) {
			super(uniqueId, displayName);
		}

		@Override
		public boolean isContainer() {
			return false;
		}

		@Override
		public boolean isTest() {
			return true;
		}
	}

}
```

**Comment by sbrannen** [comment]
FYI: I think you may be slightly confused about `request.getRootTestDescriptor()`.

That will return the `EngineDescriptor` returned from `discover()`. That's why I tracked your custom `RootTestDescriptor` separately in my updated version of your engine.

**Comment by sbrannen** [comment]
Also, you created your _container_ as a `TestTestDescriptor` when it should in fact be a `ContainerTestDescriptor`.

**Comment by mfulton26** [comment]
I was just observing that about the rootTestDescriptor. Thank you. I'm still learning as I go. :smile: 

The `TestTestDescriptor` for my container was probably from me just experimenting with it to see if returning different values for `isTest` and `isContainer` made any difference (which I didn't suspect it would and it didn't but I thought it worth a try). I must have forgotten to change it back.

Thank you for getting this updated to get me going again. I'll watch for the fix. Thank you again!

**Comment by sbrannen** [comment]
FYI: this has been resolved in `master` in 47d20a7448d441f3239b9b2ace53fe45396d9859.

**Comment by mfulton26** [comment]
This is great. It registers and executes my tests as I expect. Thank you.

I did however run into a snag but instead of appending it here I've created #368 to track it.
